plugin/hover.py

```diff
@@ -212,6 +214,10 @@ def _on_navigate(self, href: str, point: int) -> None:
                 position = {"line": row, "character": col_utf16}  # type: Position
                 r = {"start": position, "end": position}  # type: RangeLsp
                 sublime.set_timeout_async(functools.partial(session.open_uri_async, uri, r))
+        elif os.path.isabs(href) and is_file(href):
```

Comment:
This won't work when the Julia LS is running in a Docker or other remote location. You don't expect such scenarios?

Reply:
Shouldn't this be fixed in the Julia LS? There's already a handler for `file:` schemes.

Reply:
> This won't work when the Julia LS is running in a Docker or other remote location. You don't expect such scenarios?

Good to know, I did not consider this. I have very little experience with Docker, so I cannot tell if this is a realistic scenario or not.

I've looked at the language server code and saw that they even have special casing only for VS Code to use URIs with `file:` scheme (https://github.com/julia-vscode/LanguageServer.jl/blob/8d8408edb3b4bd8d1266901b9aac3f540b04fdfe/src/requests/hover.jl#L114-L116). So I've adjusted the code in the server to do the same for Sublime Text and wanted to prepare a PR there, but it still doesn't work because apparently we can't handle percent-encoding for `:` in paths (after the drive letter) - the server reports for example URIs like `file:///C%3A/Users/jwortmann/AppData/Local/Programs/Julia-1.6.1/share/julia/base/iobuffer.jl#473` in that case, but Sublime cannot open the filepath starting with `C%3A`. Iirc we had seen this problem earlier already, but I don't remember whether the percent-encoding is strictly neccessary or not. According to https://en.wikipedia.org/wiki/Percent-encoding#Reserved_characters the colon is a reserved character. But regardless of that, the percent-encoding is probably allowed, so I guess we should handle that on the client side?

Reply:
According to https://en.wikipedia.org/wiki/Uniform_Resource_Identifier the percent-encoding is required for the `:` in the path:
> Of the ASCII character set, the characters : / ? # [ ] @ are reserved for use as delimiters of the generic URI components and must be percent-encoded

We can just use `urllib.parse.unquote` to decode the URI. But ST still cannot open the file (at least for me on Windows), probably because the path begins with a `/`. But that URI with 3 `/` seems still correct and only 2 slashes would actually be invalid: [https://en.wikipedia.org/wiki/File_URI_scheme#How_many_slashes?](https://en.wikipedia.org/wiki/File_URI_scheme#How_many_slashes?)

So I guess we also need to strip the leading `/` from the path first.
How/why do `file:` URIs work for other servers with the current implementation?

Reply:
It doesn't matter what specs say when Microsoft is involved. What this code does is the authority: https://github.com/microsoft/vscode-uri

Reply:
Well, I'm mostly interested in getting the language server with ST to work, and apparently they are following the specs, but we don't support the specs here. Since the server targets VS Code in the first place, that would imply that VS Code does also support the specs, right?

Reply:
I haven't read through the RFCs but you're probably correct. The current file: scheme handling in hover.py works well for Linux and macOS, but probably doesn't work well for Windows. There is still doubt about what to do in plugin/core/url.py. Not many people dare to change it because the major contributors to this package are not using Windows.

I would have expected [urlparse](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlparse) to handle the complexity for us, but apparently it doesn't. Your proposal of using urllib.parse.unquote sounds appealing. See also this issue for some history https://github.com/sublimelsp/LSP/issues/289 on why we have this in url.py in the first place.

Reply:
BTW, after applying `unquote` I've been toying with the solution of applying `re.sub` with regex `r'^/([a-zA-Z]:)'`:

```
>>> re.sub(r'^/([a-zA-Z]:)', r'\1', '/C:/foo/bar')
'C:/foo/bar'
```

Reply:
Yeah the documentation for `urlparse` says
> The delimiters as shown above are not part of the result, except for a leading slash in the path component, which is retained if present.

which is expected on Linux/macOS, but results in an additional `/` preceding the drive letter on Windows and ST doesn't like that.

My solution would be something like this:
```python
parsed = urlparse(unquote(href))
filepath = parsed.path
if sublime.platform() == "windows" and filepath[0] == "/":
    filepath = filepath[1:]
fn = "{}:{}".format(filepath, parsed.fragment) if parsed.fragment else parsed.path
```